doc/dev/architecture.md

```diff
@@ -0,0 +1,304 @@
+
+- **Bufferization:**
+
+  - Bufferization is a process by which operations are transformed from operating on *tensors* to
+    operating on memory, represented by *memrefs* (Memory References) in MLIR. The key difference
+    between the two is that tensors behave according to value semantics, that is they cannot be
+    modified in-place. Instead, operations consume and produce new tensor values.
+
+  - In order to bufferize a program, memory has to be allocated and a buffer assigned to each
+    tensor, reusing buffers whenever possible to minimize unnecessary data copies, and eventually
+    deallocating buffers when they are no longer needed to prevent memory leaks.
+
+  - Bufferization should generally be the last step before converting to the LLVM dialect, as
+    optimizations are typically easier to implement in the tensor domain than in the memory domain.
+
+- **LLVM dialect generation:**
+
+  - As an intermediate step in the LLVM IR generation, the LLVM dialect in MLIR provides a simple
+    target for other dialects to lower to, and simplifies the conversion between MLIR and LLVM IR
+    by providing a one-to-one mapping from MLIR objects to LLVM IR objects.
+
+  - Generally, dialects will make use of the dialect conversion infrastructure to provide lowerings
+    to the LLVM dialect. The quantum dialect provides lowering patterns to QIR-style operations, and
+    the Catalyst gradient dialect can lower to device-based implementations of quantum AD or to an
+    Enzyme-based implementation for AD on classical code.
+
+- **LLVM IR generation:**
+
+  - Conversion from the LLVM dialect in MLIR to LLVM IR is handled by the `mlir-translate` tool.
+
+- **Enzyme auto-differentiation:**
+
+  - Functions that have been set-up to be differentiated via Enzyme will be transformed at this
+    stage. The seperately compiled Enzyme library is loaded into the LLVM `opt` tool to perform
+    the relevant code transformations.
+
+- **LLVM optimizations:**
+
+  - The `opt` tool can also be used to run additional LLVM passes at this point, such as
+    optimizations.
+
+  - For quantum bound programs, optimizations at this stage may see little benefit however, and are
+    generally better left out to save on compilation time.
+
+- **Native code generation:**
+
+  - The LLVM static compiler (`llc`) is invoked to perform code generation for the local target
+    architecture. A single object file is produced for the entire user program.
+
+- **Linking:**
+
+  - Using a linker available on the system, the user program is linked against the Catalyst runtime.
+    To simplify the process, a compiler will be used to drive the linking process, such as `clang`,
+    `gcc`, or `c99`.
+
+  - The shared library produced by the linking step is the output of the compilation process.
+
+
+## Runtime & Execution
+
+In progress
+
+![img](../_static/arch/runtime.svg)
+
+The Catalyst runtime essentially acts as a bridge between two public interfaces:
```

Comment:
What I mean is that the runtime bridges the C-API to the C++ API. It does this by implementing the C-API with calls to the C++ API. The C++ API is left for devices to implement. Does that make sense / do you have a suggestion to express it better? :)

Reply:
It looks clear to me now. I think it would be better to say just that: the runtime implements C-API and calls C++ API implemented by others. Btw, do you think that the direction of the arrows from the Backends to the Runtime should be reversed? (assuming they mean "uses")

Reply:
>Btw, do you think that the direction of the arrows on the picture from the Backends to the Runtime should be reversed? (assuming they mean "uses")

In the context of C4 I think they mean implement, so the direction is correct. I experimented adding it as a label but it was kind of ugly, maybe I can add it to the left-most arrow only.

Reply:
>It looks clear to me now. I think it would be better to say just that: the runtime implements C-API and calls C++ API implemented by others (in a nicer manner:)

I think I prefer the bridge formulation here, but thank you for the suggestion :)

Reply:
Added the label to the left-most arrow